File: src/services/snssai-mapping.ts
import { getCollection, handleDatabaseError } from '../db/mongodb';
import { SnssaiMapping } from '../types/db-types';
import { Snssai, PlmnId, Tai } from '../types/common-types';
import { MappingOfSnssai } from '../types/nnssf-nsselection-types';

const snssaiMatches = (s1: Snssai, s2: Snssai): boolean => {
  return s1.sst === s2.sst && s1.sd === s2.sd;
};

const plmnMatches = (p1: PlmnId, p2: PlmnId): boolean => {
  return p1.mcc === p2.mcc && p1.mnc === p2.mnc;
};

const taiMatches = (t1: Tai, t2: Tai): boolean => {
  return plmnMatches(t1.plmnId, t2.plmnId) && t1.tac === t2.tac;
};

const isMappingValidInTai = (mapping: SnssaiMapping, tai?: Tai): boolean => {
  if (!tai || !mapping.validityArea || mapping.validityArea.length === 0) {
    return true;
  }

  return mapping.validityArea.some(validTai => taiMatches(validTai, tai));
};

export const getMappingForSnssai = async (
  servingSnssai: Snssai,
  servingPlmnId: PlmnId,
  homePlmnId: PlmnId,
  tai?: Tai
): Promise<Snssai | null> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');

    const mappings = await mappingCollection.find({
      'servingSnssai.sst': servingSnssai.sst,
      'servingSnssai.sd': servingSnssai.sd,
      'servingPlmnId.mcc': servingPlmnId.mcc,
      'servingPlmnId.mnc': servingPlmnId.mnc,
      'homePlmnId.mcc': homePlmnId.mcc,
      'homePlmnId.mnc': homePlmnId.mnc
    }).toArray();

    for (const mapping of mappings) {
      if (isMappingValidInTai(mapping, tai)) {
        return mapping.homeSnssai;
      }
    }

    return null;
  } catch (error) {
    console.error('Error getting mapping for S-NSSAI:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

export const getReverseMappingForSnssai = async (
  homeSnssai: Snssai,
  servingPlmnId: PlmnId,
  homePlmnId: PlmnId,
  tai?: Tai
): Promise<Snssai | null> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');

    const mappings = await mappingCollection.find({
      'homeSnssai.sst': homeSnssai.sst,
      'homeSnssai.sd': homeSnssai.sd,
      'servingPlmnId.mcc': servingPlmnId.mcc,
      'servingPlmnId.mnc': servingPlmnId.mnc,
      'homePlmnId.mcc': homePlmnId.mcc,
      'homePlmnId.mnc': homePlmnId.mnc
    }).toArray();

    for (const mapping of mappings) {
      if (isMappingValidInTai(mapping, tai)) {
        return mapping.servingSnssai;
      }
    }

    return null;
  } catch (error) {
    console.error('Error getting reverse mapping for S-NSSAI:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

export const processMappingRequests = async (
  snssaisForMapping: Snssai[],
  servingPlmnId: PlmnId,
  homePlmnId: PlmnId,
  tai?: Tai
): Promise<MappingOfSnssai[]> => {
  const mappings: MappingOfSnssai[] = [];

  for (const servingSnssai of snssaisForMapping) {
    const homeSnssai = await getMappingForSnssai(
      servingSnssai,
      servingPlmnId,
      homePlmnId,
      tai
    );

    if (homeSnssai) {
      mappings.push({
        servingSnssai,
        homeSnssai
      });
    }
  }

  return mappings;
};

export const createSnssaiMapping = async (mapping: Omit<SnssaiMapping, 'mappingId'>): Promise<SnssaiMapping> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');

    const existingMapping = await mappingCollection.findOne({
      'servingSnssai.sst': mapping.servingSnssai.sst,
      'servingSnssai.sd': mapping.servingSnssai.sd,
      'servingPlmnId.mcc': mapping.servingPlmnId.mcc,
      'servingPlmnId.mnc': mapping.servingPlmnId.mnc,
      'homePlmnId.mcc': mapping.homePlmnId.mcc,
      'homePlmnId.mnc': mapping.homePlmnId.mnc
    });

    if (existingMapping) {
      throw new Error('S-NSSAI mapping already exists');
    }

    const mappingId = `mapping-${mapping.servingPlmnId.mcc}${mapping.servingPlmnId.mnc}-${mapping.homePlmnId.mcc}${mapping.homePlmnId.mnc}-${mapping.servingSnssai.sst}${mapping.servingSnssai.sd || ''}`;

    const newMapping: SnssaiMapping = {
      ...mapping,
      mappingId
    };

    await mappingCollection.insertOne(newMapping as any);
    return newMapping;
  } catch (error) {
    console.error('Error creating S-NSSAI mapping:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

export const getAllSnssaiMappings = async (): Promise<SnssaiMapping[]> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');
    return await mappingCollection.find({}).toArray();
  } catch (error) {
    console.error('Error getting all S-NSSAI mappings:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

export const getSnssaiMappingById = async (mappingId: string): Promise<SnssaiMapping | null> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');
    return await mappingCollection.findOne({ mappingId });
  } catch (error) {
    console.error('Error getting S-NSSAI mapping by ID:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

export const updateSnssaiMapping = async (
  mappingId: string,
  updates: Partial<Omit<SnssaiMapping, 'mappingId'>>
): Promise<SnssaiMapping | null> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');

    const result = await mappingCollection.findOneAndUpdate(
      { mappingId },
      { $set: updates },
      { returnDocument: 'after' }
    );

    return result || null;
  } catch (error) {
    console.error('Error updating S-NSSAI mapping:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

export const deleteSnssaiMapping = async (mappingId: string): Promise<boolean> => {
  try {
    const mappingCollection = getCollection<SnssaiMapping>('snssai-mappings');

    const result = await mappingCollection.deleteOne({ mappingId });
    return result.deletedCount > 0;
  } catch (error) {
    console.error('Error deleting S-NSSAI mapping:', error);
    const dbError = handleDatabaseError(error);
    throw dbError;
  }
};

File: src/services/snssai_mapping.rs
use bson::doc;
use futures::TryStreamExt;
use mongodb::Database;
use tracing::debug;

use crate::errors::{AppError, AppResult};
use crate::types::common::{PlmnId, Snssai, Tai};
use crate::types::db::SnssaiMapping;
use crate::validation::snssai_eq;

use super::allowed_nssai::tai_eq;

pub async fn map_snssai(
    serving_snssai: &Snssai,
    serving_plmn: &PlmnId,
    home_plmn: &PlmnId,
    db: &Database,
) -> AppResult<Option<Snssai>> {
    let collection = db.collection::<SnssaiMapping>("snssai_mappings");

    let cursor = collection
        .find(doc! {
            "servingSnssai.sst": serving_snssai.sst as i32,
            "servingPlmnId.mcc": &serving_plmn.mcc,
            "servingPlmnId.mnc": &serving_plmn.mnc,
            "homePlmnId.mcc": &home_plmn.mcc,
            "homePlmnId.mnc": &home_plmn.mnc,
        })
        .await
        .map_err(|e| {
            AppError::InternalServerError(format!("Failed to query snssai_mappings: {}", e))
        })?;

    let mappings: Vec<SnssaiMapping> = cursor.try_collect().await.map_err(|e| {
        AppError::InternalServerError(format!("Failed to read snssai_mappings: {}", e))
    })?;

    let result = mappings
        .iter()
        .find(|m| snssai_eq(serving_snssai, &m.serving_snssai))
        .map(|m| m.home_snssai.clone());

    debug!(
        sst = serving_snssai.sst,
        sd = ?serving_snssai.sd,
        serving_mcc = %serving_plmn.mcc,
        serving_mnc = %serving_plmn.mnc,
        home_mcc = %home_plmn.mcc,
        home_mnc = %home_plmn.mnc,
        found = result.is_some(),
        "S-NSSAI mapping lookup"
    );

    Ok(result)
}

pub async fn map_snssai_with_tai(
    serving_snssai: &Snssai,
    serving_plmn: &PlmnId,
    home_plmn: &PlmnId,
    current_tai: &Tai,
    db: &Database,
) -> AppResult<Option<Snssai>> {
    let collection = db.collection::<SnssaiMapping>("snssai_mappings");

    let cursor = collection
        .find(doc! {
            "servingSnssai.sst": serving_snssai.sst as i32,
            "servingPlmnId.mcc": &serving_plmn.mcc,
            "servingPlmnId.mnc": &serving_plmn.mnc,
            "homePlmnId.mcc": &home_plmn.mcc,
            "homePlmnId.mnc": &home_plmn.mnc,
        })
        .await
        .map_err(|e| {
            AppError::InternalServerError(format!("Failed to query snssai_mappings: {}", e))
        })?;

    let mappings: Vec<SnssaiMapping> = cursor.try_collect().await.map_err(|e| {
        AppError::InternalServerError(format!("Failed to read snssai_mappings: {}", e))
    })?;

    for mapping in &mappings {
        if !snssai_eq(serving_snssai, &mapping.serving_snssai) {
            continue;
        }

        match &mapping.validity_area {
            None => return Ok(Some(mapping.home_snssai.clone())),
            Some(tais) if tais.is_empty() => return Ok(Some(mapping.home_snssai.clone())),
            Some(tais) => {
                if tais.iter().any(|t| tai_eq(t, current_tai)) {
                    return Ok(Some(mapping.home_snssai.clone()));
                }
            }
        }
    }

    debug!(
        sst = serving_snssai.sst,
        sd = ?serving_snssai.sd,
        tac = %current_tai.tac,
        "No valid S-NSSAI mapping found for TAI"
    );

    Ok(None)
}
